src/thunderbolt/store.c

```diff
@@ -0,0 +1,465 @@
+        assert(device->uuid);
+
+        path = strjoina(store->path, "/devices/", device->uuid);
+
+        fd = open(path, O_RDONLY|O_CLOEXEC|O_NOCTTY|O_NOFOLLOW);
+        if (fd < 0)
+                return -errno;
+        if (fstat(fd, &st) < 0)
+                return -errno;
+        if (S_ISDIR(st.st_mode))
+                return -EISDIR;
+        if (!S_ISREG(st.st_mode))
+                return -ENOTTY;
+
+        f = fdopen(fd, "re");
+        if (f == NULL)
+                return -errno;
+
+        r = config_parse(NULL, path, f,
+                         NULL,
+                         config_item_table_lookup, items,
+                         true, true, false, device);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to parse %s: %m", device->uuid);
+
+        return 0;
+}
+
+int tb_store_device_load(TbStore *store, const char *uuid, TbDevice **device) {
+        TbDevice *d = NULL;
+        int r;
+
+        assert(store);
+        assert(device);
+        assert(uuid);
+
+        d = new0(TbDevice, 1);
+        if (!d) {
+                r = -ENOMEM;
+                goto out;
+        }
+
+        d->uuid = strdup(uuid);
+        if (d->uuid == NULL) {
+                r = -ENOMEM;
+                tb_device_free(&d);
+                goto out;
+        }
+
+        r = tb_store_parse_device(store, d);
+        if (r < 0) {
+                tb_device_free(&d);
+        }
+
+ out:
+        *device = d;
+        return r;
+}
+
+
+static int store_efivars_get_auth(const char *uuid, Auth *ret) {
+        _cleanup_free_ void *var = NULL;
+        sd_id128_t id;
+        size_t l;
+        int r;
+
+        assert(ret);
+        assert(uuid);
+
+        if (sd_id128_from_string(uuid, &id) < 0) {
+                return -EINVAL;
+        }
+
+        r = efi_get_variable(id, "Thunderbolt", NULL, &var, &l);
+        if (r < 0)
+                return r;
+
+        ret->store = STORE_EFIVARS;
+        if (l == 1) {
+                return safe_atoi(var, &ret->level);
+        } else if (l == KEY_CHARS) {
+                ret->level = AUTH_KEY;
+                ret->key = (char *) var;
+                var = NULL;
+                return 0;
+        }
+
+        /* should not happen, because only we write it */
+        return -EIO;
+}
+
+int store_get_auth(TbStore *store, const char *uuid, Auth *ret) {
+        _cleanup_free_ char *p = NULL;
+        struct stat st;
+        char *path;
+        int r;
+
+        if (in_initrd())
+                return store_efivars_get_auth(uuid, ret);
+
+        path = strjoina(store->path, "/authorization/", uuid);
+
+        r = lstat(path, &st);
+        if (r < 0 && errno == ENOENT) {
+                ret->level = AUTH_MISSING;
+                return 0;
+        }
+        if (r < 0)
+                return -errno;
+        if (S_ISREG(st.st_mode)) {
+                _cleanup_free_ char *l = NULL;
+
+                r = read_one_line_file(path, &l);
```

Comment:
hmm, this has a "racy" feel to it: doing lstat() first, and then later on reading the file might trigger a race condition if this file is changed by something else in between. Please always open first, then use fstat() to validate, so that we always operate on the same file object